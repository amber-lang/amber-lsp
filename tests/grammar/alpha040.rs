use std::fs::read_to_string;

use chumsky::error::Rich;
use insta::assert_debug_snapshot;

use amber_lsp::grammar::{
    alpha040::{lexer::Token, AmberCompiler, GlobalStatement, Spanned},
    LSPAnalysis, ParserResponse,
};

fn tokenize(input: &str) -> Vec<Spanned<Token>> {
    AmberCompiler::new().tokenize(input)
}

fn parse(
    tokens: &[Spanned<Token>],
) -> (Option<Vec<Spanned<GlobalStatement>>>, Vec<Rich<'_, String>>) {
    let ParserResponse {
        ast,
        errors,
        semantic_tokens: _,
    } = AmberCompiler::new().parse(tokens);

    let ast = match ast {
        amber_lsp::grammar::Grammar::Alpha040(ast) => ast,
        _ => panic!("Unexpected AST"),
    };

    (ast, errors)
}

fn parse_unwrap(tokens: &[Spanned<Token>]) -> Vec<Spanned<GlobalStatement>> {
    let (ast, errors) = parse(tokens);
    if !errors.is_empty() {
        panic!("Errors: {:?}", errors);
    }
    ast.unwrap()
}

#[test]
fn test_stdlib_array() {
    let stdlib = read_to_string("resources/alpha040/std/array.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_date() {
    let stdlib = read_to_string("resources/alpha040/std/date.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_env() {
    let stdlib = read_to_string("resources/alpha040/std/env.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_fs() {
    let stdlib = read_to_string("resources/alpha040/std/fs.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_http() {
    let stdlib = read_to_string("resources/alpha040/std/http.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_math() {
    let stdlib = read_to_string("resources/alpha040/std/math.ab").unwrap();

    assert_debug_snapshot!(parse_unwrap(&tokenize(&stdlib)));
}

#[test]
fn test_stdlib_text() {
    let stdlib = read_to_string("resources/alpha040/std/text.ab").unwrap();

    let tokens = tokenize(&stdlib);

    assert_debug_snapshot!(tokens);

    assert_debug_snapshot!(parse_unwrap(&tokens));
}

#[test]
fn test_unfinished_function_call() {
    let input = "
    import { array_contains } from \"std/array\"
    
    let x = [1, 2, 3]
    let y = 2
    
    echo array_contains(x, y)
    
    let line = 213
    
    for idx, line in lines(\"text.txt\") {
      echo line
    }
    
    // fun foo(x: Num, y: Text) {
    
    // }
    
    fun foo(x: Num, y: Text) {
    }
    
    // fun abc() {}
    
    array_contains([1, 2, 3],)
    ";

    let tokens = tokenize(input);
    assert_debug_snapshot!(tokens);

    let result = parse(&tokens);

    assert_debug_snapshot!(result);
}
