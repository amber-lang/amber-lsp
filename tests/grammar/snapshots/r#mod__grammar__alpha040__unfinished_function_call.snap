---
source: tests/grammar/alpha040.rs
expression: tokens
---
[
    (
        Token(
            "import",
        ),
        5..11,
    ),
    (
        Token(
            "{",
        ),
        12..13,
    ),
    (
        Token(
            "array_contains",
        ),
        14..28,
    ),
    (
        Token(
            "}",
        ),
        29..30,
    ),
    (
        Token(
            "from",
        ),
        31..35,
    ),
    (
        Token(
            "\"",
        ),
        36..37,
    ),
    (
        Token(
            "std",
        ),
        37..40,
    ),
    (
        Token(
            "/",
        ),
        40..41,
    ),
    (
        Token(
            "array",
        ),
        41..46,
    ),
    (
        Token(
            "\"",
        ),
        46..47,
    ),
    (
        Token(
            "let",
        ),
        57..60,
    ),
    (
        Token(
            "x",
        ),
        61..62,
    ),
    (
        Token(
            "=",
        ),
        63..64,
    ),
    (
        Token(
            "[",
        ),
        65..66,
    ),
    (
        Token(
            "1",
        ),
        66..67,
    ),
    (
        Token(
            ",",
        ),
        67..68,
    ),
    (
        Token(
            "2",
        ),
        69..70,
    ),
    (
        Token(
            ",",
        ),
        70..71,
    ),
    (
        Token(
            "3",
        ),
        72..73,
    ),
    (
        Token(
            "]",
        ),
        73..74,
    ),
    (
        Token(
            "let",
        ),
        79..82,
    ),
    (
        Token(
            "y",
        ),
        83..84,
    ),
    (
        Token(
            "=",
        ),
        85..86,
    ),
    (
        Token(
            "2",
        ),
        87..88,
    ),
    (
        Token(
            "echo",
        ),
        98..102,
    ),
    (
        Token(
            "array_contains",
        ),
        103..117,
    ),
    (
        Token(
            "(",
        ),
        117..118,
    ),
    (
        Token(
            "x",
        ),
        118..119,
    ),
    (
        Token(
            ",",
        ),
        119..120,
    ),
    (
        Token(
            "y",
        ),
        121..122,
    ),
    (
        Token(
            ")",
        ),
        122..123,
    ),
    (
        Token(
            "let",
        ),
        133..136,
    ),
    (
        Token(
            "line",
        ),
        137..141,
    ),
    (
        Token(
            "=",
        ),
        142..143,
    ),
    (
        Token(
            "213",
        ),
        144..147,
    ),
    (
        Token(
            "for",
        ),
        157..160,
    ),
    (
        Token(
            "idx",
        ),
        161..164,
    ),
    (
        Token(
            ",",
        ),
        164..165,
    ),
    (
        Token(
            "line",
        ),
        166..170,
    ),
    (
        Token(
            "in",
        ),
        171..173,
    ),
    (
        Token(
            "lines",
        ),
        174..179,
    ),
    (
        Token(
            "(",
        ),
        179..180,
    ),
    (
        Token(
            "\"",
        ),
        180..181,
    ),
    (
        Token(
            "text",
        ),
        181..185,
    ),
    (
        Token(
            ".",
        ),
        185..186,
    ),
    (
        Token(
            "txt",
        ),
        186..189,
    ),
    (
        Token(
            "\"",
        ),
        189..190,
    ),
    (
        Token(
            ")",
        ),
        190..191,
    ),
    (
        Token(
            "{",
        ),
        192..193,
    ),
    (
        Token(
            "echo",
        ),
        200..204,
    ),
    (
        Token(
            "line",
        ),
        205..209,
    ),
    (
        Token(
            "}",
        ),
        214..215,
    ),
    (
        Token(
            "// fun foo(x: Num, y: Text) {\n",
        ),
        225..255,
    ),
    (
        Token(
            "// }\n",
        ),
        264..269,
    ),
    (
        Token(
            "fun",
        ),
        278..281,
    ),
    (
        Token(
            "foo",
        ),
        282..285,
    ),
    (
        Token(
            "(",
        ),
        285..286,
    ),
    (
        Token(
            "x",
        ),
        286..287,
    ),
    (
        Token(
            ":",
        ),
        287..288,
    ),
    (
        Token(
            "Num",
        ),
        289..292,
    ),
    (
        Token(
            ",",
        ),
        292..293,
    ),
    (
        Token(
            "y",
        ),
        294..295,
    ),
    (
        Token(
            ":",
        ),
        295..296,
    ),
    (
        Token(
            "Text",
        ),
        297..301,
    ),
    (
        Token(
            ")",
        ),
        301..302,
    ),
    (
        Token(
            "{",
        ),
        303..304,
    ),
    (
        Token(
            "}",
        ),
        309..310,
    ),
    (
        Token(
            "// fun abc() {}\n",
        ),
        320..336,
    ),
    (
        Token(
            "array_contains",
        ),
        345..359,
    ),
    (
        Token(
            "(",
        ),
        359..360,
    ),
    (
        Token(
            "[",
        ),
        360..361,
    ),
    (
        Token(
            "1",
        ),
        361..362,
    ),
    (
        Token(
            ",",
        ),
        362..363,
    ),
    (
        Token(
            "2",
        ),
        364..365,
    ),
    (
        Token(
            ",",
        ),
        365..366,
    ),
    (
        Token(
            "3",
        ),
        367..368,
    ),
    (
        Token(
            "]",
        ),
        368..369,
    ),
    (
        Token(
            ",",
        ),
        369..370,
    ),
    (
        Token(
            ")",
        ),
        370..371,
    ),
]
